64_MergeSort/main.cpp

```diff
@@ -0,0 +1,368 @@
+#include "nbl/application_templates/MonoDeviceApplication.hpp"
+#include "nbl/application_templates/MonoAssetManagerAndBuiltinResourceApplication.hpp"
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace asset;
+using namespace video;
+
+// A high level overview of the algorithm implementation (merge sort).
+// This will be a X pass algorithm, where X is log2(number_of_elements).
+// In each phase, you have to 'merge' 2 sorted list of elements into a single list.
+// Each of these 'lists' will have thier number of elements roughly doubled in each phase (after the merge step of previous phase).
+class MergeSortApp final : public nbl::application_templates::MonoDeviceApplication, public application_templates::MonoAssetManagerAndBuiltinResourceApplication
+{
+	using device_base_t = application_templates::MonoDeviceApplication;
+	using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
+
+public:
+	MergeSortApp(const path& _localInputCWD, const path& _localOutputCWD, const path& _sharedInputCWD, const path& _sharedOutputCWD) :
+		system::IApplicationFramework(_localInputCWD, _localOutputCWD, _sharedInputCWD, _sharedOutputCWD) {}
+
+	bool onAppInitialized(smart_refctd_ptr<ISystem>&& system) override
+	{
+		// Initialize base classes.
+		if (!device_base_t::onAppInitialized(smart_refctd_ptr(system)))
+		{
+			return false;
+		}
+		if (!asset_base_t::onAppInitialized(std::move(system)))
+		{
+			return false;
+		}
+
+		// Get the maxinum bytes of shared memory we can use.
+		// In case the number of elements shared memory array can hold exceeds the limit, then switch to fetching data from global memory directly (for as few memory elements as possible).
+		const auto physicalDeviceLimits = m_device->getPhysicalDevice()->getLimits();
+		const uint32_t maxNumberOfArrayElementsSharedMemoryCanHold = physicalDeviceLimits.maxComputeSharedMemorySize / sizeof(int);
+
+		// Compute and input related constants.
+		static constexpr uint32_t WorkgroupSize = 64;
+		static constexpr uint32_t NumberOfElementsToSort = 1024 * 633u;
+		static const uint32_t WorkgroupCount = (uint32_t)ceil(NumberOfElementsToSort / (float)WorkgroupSize);
```

Comment:
just stop it, stop using floats for integer math

Reply:
Also use virtual (persistent) workgroups, so you `max()` the workgroup count against a device specific limit (`SPhysicalDeviceLimits` has a helper func for this)